### ISSUE TITLE
[Feature request] Implement FSRS because SM2 is flawed

### ISSUE THREAD

**Issue by v-Nyo** [l3kn/org-fc#129]
FSRS:
https://github.com/open-spaced-repetition

Detailed comparison to SM2 (Anki):
https://www.youtube.com/watch?v=OqRLqVRyIzc

I also experience that SM2 shows cards far too often in the beginning and is far to optimistic after a while.

An Implementation of FSRS for Emacs would be awesome!

EDIT:
Just noticed in you activity overview  that you already had some discussions about FSRS not sure if you are interested in implementing it in org-fc.

**Comment by cashpw** [comment]
> Just noticed in you activity overview that you already had some discussions about FSRS not sure if you are interested in implementing it in org-fc.

Please add a link. I can't find this discussion.

**Comment by l3kn** [comment]
Some of the last changes made to org-fc (support for multiple algorithms and refactoring the dashboard) were intended to ease the way for adding FSRS.

There is an experimental implementation of FSRS v4 that I've been using for a while but there are some problems that need to be solved before I'd want to officially add that to org-fc.

I think one key part of FSRS is that it allows optimizing the parameters for different collections of cards. In Anki this is easy because cards are grouped into decks. With the tag based approach of org-fc this is a bit harder and will require an interface between org-fc and either the python or the rust optimizer of FSRS.

To account for changes in the FSRS parameters or updates of the underlying algorithm (v4.5 and the recent v5),
we'll need a way to re-calculate the difficulty/stability/due-date of cards (I thinks it's fine to do this after a review instead of updating all cards in the system at once).

A related open question is how to convert existing cards to FSRS, as org-fc does not track new/(re)learning/review states like Anki and those are used by FSRS. I have tried to evaluate different ways of getting around this (e.g. by considering the time between reviews and treating everything shorter than one day as a learning phase) but it's not clear what's the best approach.

Version 5 of FSRS includes a model of short-term recall but I have not yet looked into how that might solve this issue.

To summarize: If you're interested in trying FSRS (with default parameters) in org-fc, I can provide a somewhat working implementation of v4. A more feature rich and stable integration will require a significant amount of work.

**Comment by cashpw** [comment]
I don't have time to do a full implementation but perhaps this can start things off:

```elisp

(defconst org-fc-algo-fsrs-6-review-card-script-path
  "/home/cashpw/proj/org-fc/org-fc-algo-fsrs-6-review-card.py"
  "Path to review-card.py.")

(defcustom org-fc-algo-fsrs-6-parameters
  '(0.2172
    1.1771
    3.2602
    16.1507
    7.0114
    0.57
    2.0966
    0.0069
    1.5261
    0.112
    1.0178
    1.849
    0.1133
    0.3127
    2.2934
    0.2191
    3.0004
    0.7536
    0.3332
    0.1437
    0.2)
  "The model weights of the FSRS scheduler."
  :type '(repeat float)
  :group 'org-fc)

(defcustom org-fc-algo-fsrs-6-desired-retention 0.9
  "The desired retention rate of cards scheduled with the scheduler."
  :type 'float
  :group 'org-fc)

(defcustom org-fc-algo-fsrs-6-learning-steps
  (list (seconds-to-time 60) (seconds-to-time 600))
  "Small time intervals that schedule cards in the Learning state."
  :type '(repeat sexp)
  :group 'org-fc)

(defcustom org-fc-algo-fsrs-6-relearning-steps (list (seconds-to-time 600))
  "Small time intervals that schedule cards in the Relearning state."
  :type '(repeat sexp)
  :group 'org-fc)

(defcustom org-fc-algo-fsrs-6-maximum-interval (* 100 365)
  "The maximum number of days a Review-state card can be scheduled into the future"
  :type 'number
  :group 'org-fc)

(defcustom org-fc-algo-fsrs-6-enable-fuzzing t
  "Whether to apply a small amount of random 'fuzz' to calculated intervals."
  :type 'boolean
  :group 'org-fc)

(defun org-fc-algo-fsrs-6-review-card
    (id state step stability difficulty due last-review rating)
  "Return results of review-card script."
  (shell-command-to-string
   (concat
    "python " org-fc-algo-fsrs-6-review-card-script-path
    (format " --scheduler_parameters %s"
            (string-join (mapcar
                          #'number-to-string org-fc-algo-fsrs-6-parameters)
                         " "))
    (format " --scheduler_desired_retention %s"
            org-fc-algo-fsrs-6-desired-retention)
    (format " --scheduler_learning_steps %s"
            (string-join (mapcar
                          (lambda (step)
                            (number-to-string
                             (truncate (time-to-seconds step))))
                          org-fc-algo-fsrs-6-learning-steps)
                         " "))
    (format " --scheduler_relearning_steps %s"
            (string-join (mapcar
                          (lambda (step)
                            (number-to-string
                             (truncate (time-to-seconds step))))
                          org-fc-algo-fsrs-6-relearning-steps)
                         " "))
    (format " --scheduler_maximum_interval %s"
            org-fc-algo-fsrs-6-maximum-interval)
    (when org-fc-algo-fsrs-6-enable-fuzzing
      " --scheduler_enable_fuzzing")
    (format " --card_id %s" id)
    (when state
      (format " --card_state %s" state))
    (when step
      (format " --card_step %s" step))
    (when stability
      (format " --card_stability %s" stability))
    (when difficulty
      (format " --card_difficulty %s" difficulty))
    (when due
      (format " --card_due %s" (truncate (time-to-seconds due))))
    (when last-review
      (format " --card_last_review %s"
              (truncate (time-to-seconds last-review))))
    (format " --card_rating %s" rating))))

(defun org-fc-algo-fsrs-6-float-to-time (float-time)
  "Convert a floating-point Unix timestamp FLOAT-TIME to an Emacs time value."
  (let* ((seconds (truncate float-time))
         (fraction (- float-time seconds))
         (time-val (seconds-to-time seconds))
         ;; Convert the fractional part to total microseconds.
         (total-micros (* fraction 1000000.0))
         (micro (truncate total-micros))
         ;; Use the remainder to calculate picoseconds.
         (pico (truncate (* (- total-micros micro) 1000000.0))))
    ;; Update the time value list with the microsecond and picosecond values.
    (setf (nth 2 time-val) micro)
    (setf (nth 3 time-val) pico)
    time-val))

(defun org-fc-algo-fsrs-6-parse-review-card-output (output)
  "Parse the OUTPUT of `org-fc-algo-fsrs-6-review-card-script-path'."
  (cl-destructuring-bind
      (id state step stability difficulty due last-review)
      (split-string output "," 'omit-nulls)
    (list
     :id (string-to-number id)
     :state (string-to-number state)
     :step (string-to-number step)
     :stability (string-to-number stability)
     :difficulty (string-to-number difficulty)
     :due
     (org-fc-algo-fsrs-6-float-to-time (string-to-number due))
     :last-review
     (org-fc-algo-fsrs-6-float-to-time (string-to-number last-reviewed)))))

(defclass org-fc-algo-fsrs-6 (eieio-singleton org-fc-algo) ())

(cl-defmethod org-fc-algo-headers ((_algo org-fc-algo-fsrs-6))
  '(position state step stability difficulty due last-review))

(cl-defmethod org-fc-algo-initial-review-data ((_algo org-fc-algo-fsrs-6) name)
  "Initial FSRS_6 review data for position NAME."
  (list
   'position
   name
   'state
   nil
   'step
   nil
   'stability
   nil
   'difficulty
   nil
   'due
   nil
   'last-review
   nil))

(cl-defmethod org-fc-algo-next-review-data
  ((_algo org-fc-algo-fsrs-6) current rating)
  "Calculate the next parameters, given the CURRENT parameters and a RATING."
  ((org-fc-algo-fsrs-6-parse-review-card-output
    (org-fc-algo-fsrs-6-review-card
     (plist-get current :id)
     (plist-get current :state)
     (plist-get current :step)
     (plist-get current :stability)
     (plist-get current :difficulty)
     (plist-get current :due)
     (plist-get current :last-review)))))
```

```python
#!/usr/bin/env python3

from fsrs import Scheduler, Card, Rating, ReviewLog, State
from datetime import datetime, timedelta, timezone

import argparse
import json

parser = argparse.ArgumentParser(
    description="Accepts scheduler and card parameters from the command line."
)

parser.add_argument('--scheduler_parameters', nargs='+', type=float, required=True)
parser.add_argument('--scheduler_desired_retention', type=float, required=True)
parser.add_argument('--scheduler_learning_steps', nargs='+', type=int, required=True)
parser.add_argument('--scheduler_relearning_steps', nargs='+', type=int, required=True)
parser.add_argument('--scheduler_maximum_interval', type=int, required=True)
parser.add_argument('--scheduler_enable_fuzzing', action='store_true')
parser.add_argument('--card_id', type=int, required=True)
parser.add_argument('--card_state', type=int)
parser.add_argument('--card_step', type=int)
parser.add_argument('--card_stability', type=float)
parser.add_argument('--card_difficulty', type=float)
parser.add_argument('--card_due', type=int)
parser.add_argument('--card_last_review', type=int)
parser.add_argument('--card_rating', type=int, required=True)
args = parser.parse_args()

def steps_to_timedelta(seconds: int):
    timedelta(seo)

# note: the following arguments are also the defaults
scheduler = Scheduler(
    parameters = args.scheduler_parameters,
    desired_retention = args.scheduler_desired_retention,
    learning_steps = map(lambda seconds: timedelta(seconds=seconds), args.scheduler_learning_steps),
    relearning_steps = map(lambda seconds: timedelta(seconds=seconds), args.scheduler_relearning_steps),
    maximum_interval = args.scheduler_maximum_interval,
    enable_fuzzing = args.scheduler_enable_fuzzing,
)
card = Card(
    card_id = args.card_id,
    state = State(args.card_state) if args.card_state else State.Learning,
    step = args.card_step,
    stability = args.card_stability,
    difficulty = args.card_difficulty,
    due = datetime.fromtimestamp(args.card_due, tz=timezone.utc) if args.card_due else None,
    last_review = datetime.fromtimestamp(args.card_last_review, tz=timezone.utc) if args.card_last_review else None,
)

card, _ = scheduler.review_card(card, Rating(args.card_rating))

# "Return" to Emacs as output string
print(f"{card.card_id},{card.state},{card.step},{card.stability},{card.difficulty},{card.due.timestamp()},{card.last_review.timestamp()}")
```

```elisp
;; Reviewing a new position with an ID of 1234
(org-fc-algo-fsrs-6-parse-review-card-output
  (org-fc-algo-fsrs-6-review-card 1234 nil nil nil nil nil nil 3))
```

Next steps:

1. Figure out how to handle the card IDs. `org-fc` uses `org`'s UUIDs whereas `fsrs` requires integers
2. Complete the implementation of `org-fc-algo-fsrs-6`

**Comment by l3kn** [comment]
I think the optimizer works on (card_id, date, rating, duration) entries which could be computed from the existing review log format. In that case users could also optimize FSRS based on non-fsrs review. 

Optimization might be the most complex aspect of this, or at least something for which I never figured out a good approach.

We'd need a way to set different scheduler parameters for different kinds of cards (this might be possible using similar filters as the review-contexts currently implemented), collect the review logs for just these cards or positions and then update the review data of all affected cards.

A compromise might be to only update cards on their next rating, but then we need to keep track of which scheduler settings (desired retention & parameters) where used for each position.

**Comment by l3kn** [comment]
What's missing to finish this? Just the function for writing review logs (I'd just use the same as SM2 and set box, interval and ease to null)?

The python file could live in a new `python/` folder in the repo so we can keep it aligned with the emacs interface.

If you're fine with this, I can try to fit this together somehow. The current emacs files all have a copyright line with a name and a "Author: name (email)" line. What information should I use there for you?

**Comment by cashpw** [comment]
> If you're fine with this, I can try to fit this together somehow. The current emacs files all have a copyright line with a name and a "Author: name (email)" line. What information should I use there for you?

Sounds good to me!

```
Author: Cash Prokop-Weaver (cash@cashpw.com)
```

**Comment by l3kn** [comment]
In https://github.com/l3kn/org-fc/pull/137 I've made some changes to how card & scheduler data is passed
because I'd in the future I'd like to have the ability to pass data for multiple cards at once or to start the python CLI once and make multiple calls to it.

I think it should be solid enough to use with newly created flashcards but before I test it on my collection, I'd like to add some way of converting older cards to FSRS.

**Comment by cashpw** [comment]
This is fantastic! Thank you for carrying it forward from my prototype. 

> I'd like to add some way of converting older cards to FSRS.

Agreed. I'm excited to convert my collection to FSRS at some point here.

**Comment by l3kn** [comment]
I've migrated all my cards to FSRS and completed a few hundred reviews without issues so FSRS is now included in org-fc.
It requires some extra [setup steps](https://github.com/l3kn/org-fc/blob/main/docs/fsrs.org).

I'll close this for now. Please let me know if you run into any issues with it or need help migrating a large collection of files.
